File: py-orchestrator/botconfig.py
import json
from pathlib import Path
from typing import List, Optional
from pydantic import BaseModel, Field, field_validator
from rich.console import Console

console = Console()
CONFIG_FILE = Path("../config/bots_config.json")


class BotEntry(BaseModel):
    name: str
    path: str
    repo_url: str = Field(..., alias="repo_url")
    enabled: bool
    type: str  # "python" or "javascript"

    @property
    def is_bot(self) -> bool:
        """Property untuk menentukan apakah ini bot (bukan tool)."""
        return "/privatekey/" in self.path or "/token/" in self.path

    class Config:
        populate_by_name = True


class BotConfig(BaseModel):
    bots_and_tools: List[BotEntry] = Field(..., alias="bots_and_tools")

    @classmethod
    def load(cls) -> Optional["BotConfig"]:
        if not CONFIG_FILE.exists():
            console.print(f"[red]Error: File konfig '{CONFIG_FILE}' tidak ditemukan.[/]")
            return None
        try:
            with open(CONFIG_FILE, "r") as f:
                data = json.load(f)
            return cls.model_validate(data)
        except json.JSONDecodeError as e:
            console.print(f"[red]Error parsing {CONFIG_FILE}: {e}[/]")
            return None
        except Exception as e:
            console.print(f"[red]Error memuat config: {e}[/]")
            return None

# Opsi untuk menu "Back"
BACK_OPTION = BotEntry(
    name="[[Back]] Kembali", path="", repo_url="", enabled=False, type="SYSTEM"
)

File: orchestrator-tui/BotConfig.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Spectre.Console;

namespace Orchestrator;

public class BotConfig
{
    private static readonly string ProjectRoot = GetProjectRoot();
    private static readonly string ConfigFile = Path.Combine(ProjectRoot, "config", "bots_config.json");
    private static readonly string LocalBotRoot = @"D:\SC";

    private static string GetProjectRoot()
    {
        var currentDir = new DirectoryInfo(AppContext.BaseDirectory);
        
        while (currentDir != null)
        {
            var configDir = Path.Combine(currentDir.FullName, "config");
            var gitignore = Path.Combine(currentDir.FullName, ".gitignore");
            
            if (Directory.Exists(configDir) && File.Exists(gitignore))
            {
                return currentDir.FullName;
            }
            
            currentDir = currentDir.Parent;
        }
        
        return Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", ".."));
    }

    [JsonPropertyName("bots_and_tools")]
    public List<BotEntry> BotsAndTools { get; set; } = new();

    public static BotConfig? Load()
    {
        if (!File.Exists(ConfigFile))
        {
            AnsiConsole.MarkupLine($"[red]Error: File konfig '{ConfigFile}' tidak ditemukan.[/]");
            return null;
        }
        try
        {
            var json = File.ReadAllText(ConfigFile);
            return JsonSerializer.Deserialize<BotConfig>(json, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            });
        }
        catch (JsonException ex)
        {
             AnsiConsole.MarkupLine($"[red]Error parsing {ConfigFile}: {ex.Message}[/]");
             return null;
        }
         catch (Exception ex)
        {
             AnsiConsole.MarkupLine($"[red]Error reading {ConfigFile}: {ex.Message}[/]");
             return null;
        }
    }
    
    public static string GetLocalBotPath(string configPath)
    {
        configPath = configPath.Replace('/', '\\');
        var botName = Path.GetFileName(configPath).ToLowerInvariant();
        
        if (configPath.Contains("privatekey", StringComparison.OrdinalIgnoreCase))
        {
            return Path.Combine(LocalBotRoot, "PrivateKey", botName);
        }
        else if (configPath.Contains("token", StringComparison.OrdinalIgnoreCase))
        {
            return Path.Combine(LocalBotRoot, "Token", botName);
        }
        else
        {
            AnsiConsole.MarkupLine($"[dim]   Path non-standar: {configPath}. Ditaro di root D:\\SC[/]");
            return Path.Combine(LocalBotRoot, botName);
        }
    }
}

public class BotEntry
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("path")]
    public string Path { get; set; } = string.Empty;

    [JsonPropertyName("repo_url")]
    public string RepoUrl { get; set; } = string.Empty;
    
    [JsonPropertyName("enabled")]
    public bool Enabled { get; set; } = false;

    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;
    
    [JsonIgnore]
    public bool IsBot => Path.Contains("/privatekey/") || Path.Contains("/token/");
}
